packages/core/src/2d/text/TextUtils.ts

```diff
@@ -0,0 +1,413 @@
+
+    canvas.width = width;
+    canvas.height = height;
+
+    context.font = font;
+    context.fillStyle = "#000";
+    context.clearRect(0, 0, width, height);
+    context.textBaseline = "alphabetic";
+    context.fillStyle = "#f00";
+    context.fillText(testStr, 0, baseline);
+
+    const imgData = context.getImageData(0, 0, width, height).data;
+    const lineDataCount = width * 4;
+    let flag = false;
+    let i = 0;
+    let startIndex = 0;
+
+    for (i = 0; i < baseline; ++i) {
+      startIndex = i * lineDataCount;
+      for (let j = 0; j < lineDataCount; j += 4) {
+        if (imgData[startIndex + j] !== 0) {
+          flag = true;
+          break;
+        }
+      }
+      if (flag) {
+        break;
+      }
+    }
+
+    const ascent = baseline - i;
+    flag = false;
+
+    for (i = height - 1; i >= baseline; --i) {
+      startIndex = i * lineDataCount;
+      for (let j = 0; j < lineDataCount; j += 4) {
+        if (imgData[startIndex + j] !== 0) {
+          flag = true;
+          break;
+        }
+      }
+      if (flag) {
+        break;
+      }
+    }
+
+    const descent = i - baseline + 1;
+    fontSize = ascent + descent;
+    fontSizeCache[font] = fontSize;
+    return fontSize;
+  }
+
+  /**
+   * Measure the text.
+   * @param text - rendering string
+   * @param originWidth - the width of the TextRenderer
+   * @param originHeight - the height of the TextRenderer
+   * @param lineSpacing - the space between two lines
+   * @param enableWrapping - whether wrap text to next line when exceeds the width of the container
+   * @param overflowMode - the overflow mode
+   * @param fontStr - the font string
+   * @returns the TextMetrics object
+   */
+  public static measureText(
+    text: string,
+    originWidth: number,
+    originHeight: number,
+    lineSpacing: number,
+    enableWrapping: boolean,
+    overflowMode: OverflowMode,
+    fontStr: string
+  ): TextMetrics {
+    const { _pixelsPerUnit } = TextUtils;
+    const fontSize = TextUtils.measureFont(fontStr);
+    const textContext = TextUtils.textContext();
+    const { context } = textContext;
+    context.font = fontStr;
+    const lines = TextUtils._wordWrap(text, originWidth, enableWrapping, fontStr);
+    const linesLen = lines.length;
+    const lineWidths = new Array<number>();
+    const lineHeight = fontSize + lineSpacing * _pixelsPerUnit;
+    // Calculate max width of all lines.
+    let maxLineWidth = 0;
+    for (let i = 0; i < linesLen; ++i) {
+      const width = Math.ceil(context.measureText(lines[i]).width);
+      if (width > maxLineWidth) {
+        maxLineWidth = width;
+      }
+      lineWidths.push(width);
+    }
+
+    // Reset width and height.
+    const width = Math.min(maxLineWidth, TextUtils._maxWidth);
+    let height = originHeight * _pixelsPerUnit;
+    if (overflowMode === OverflowMode.Overflow) {
+      height = Math.min(lineHeight * linesLen, TextUtils._maxHeight);
```

Comment:
Why
`lineHeight * linesLen`
instead of
`fontSize * linesLen  + lineSpacing * _pixelsPerUnit * （linesLen - 1）`

Reply:
lineHeight has included lineSpacing * _pixelsPerUnit. And this is a simplified way to calculate height